### PR TITLE
WPGatsby Stability

### PR DESCRIPTION
**Pull request by jasonbahl**
This Pull Request works toward getting the plugin more stable by doing the following:

- **TESTS:** This PR adds a LOT of tests for the Action Monitor. 
- **Refactored Action Monitors:** Instead of one file that tries to track all actions, we now have many files with their own Monitor classes dedicated to tracking specific actions of specific types of data.
- **JWT Secret Set once:** Now, instead of re-generating the JWT Secret until WPGatsby settings are saved, this generates it only if it's not yet been saved once, and saves it. 

This PR addresses the following issues: 

- #70: When field groups are saved using ACF Field Group GUI, a "Diff Schema" action is triggered
- #58: A "Refetch All" action is available and is used when Permalinks are changed
- #57: Term meta is now properly tracked when changed
- #56: Custom post types (all post types that are public and show_in_graphql) are now tracked when they are moved from publish to trash and vis-versa
- #41: Codeception tests are now in place
- #38: Many core WordPress options have been added to an allow-list and trigger a general `NON_NODE_ROOT_FIELDS` action. A few specific actions trigger specific actions for specific nodes. For example, changing the home_page triggers an update for the new page and the old page being changed as the URI is now different. 
- #26: Posts that transition from future->publish now trigger an action (ensuring WordPress cron is triggered for WordPress sites using Gatsby front-ends might need more thought still though. . .)
- #17: Meta is now tracked for Posts, Terms and Users (comments are not currently tracked at the moment)
- #15: Saving permalinks triggers a `REFETCH_ALL` Action
- #7: Gatsby JWT Secret is now generated once and saved immediately and not generated again
- #6: Gatsby now tracks only post_types (and taxonomies)  that are set to be both `public` and `show_in_graphql` and there are filters to override as needed.

### CODE REVIEW

**Review by TylerBarnes — CHANGES_REQUESTED** [review]
Currently reviewing this :) I'll post things here as I find them.

Thing 1:

The user isn't updated when saving a post anymore. We need the user node to update so it has the correct list of child pages/posts/cpts attached to it.

On the first save of a post I saw that it tried to delete a post as a user:

Screen Shot 2020-12-09 at 4 06 07 PM

I'm not sure where "A published post" came from because I don't have a post or user called that

On further updates I don't see that anymore but I don't see the author being updated either:

Screen Shot 2020-12-09 at 4 06 47 PM

Thing 2:

The title is displayed in the terminal output so we want to adjust the action monitor titles to account for that for all the different node types:

Screen Shot 2020-12-09 at 4 18 42 PM

Ideally it'd still say "create page New page 2" instead of "create page CREATE New page 2". Same for other types. Generally the action monitor title should be the title of the thing being changed.

We could change it on the source plugin side so it shows the action monitor title instead of building a string out of the action type, name, and type, but that would be a breaking change 🤔 it's probably easier to keep the action monitor titles the same as the thing they're tracking.

For any reference fields it might be nicer to return null instead of "none" to make it more gql-like :)

Screen Shot 2020-12-09 at 4 25 58 PM

Holy crap, updating the homepage works!! 💜  🔥 🔥 that's so awesome!
And settings! 🚀 🚀 🚀 

The post type stuff is amazing 😱 this is all amazing!

**Comment by TylerBarnes** [comment]
Oo 2 more small things. 1: I think we need to merge master in here. I merged it locally but I'm not sure if I resolved the conflict properly. 2: Could you add an `## Upcoming` section to the changelog.md so it'll be easier to release once we merge. Thanks!

**Review by acao — APPROVED** [review]
tested these changes locally with the latest source plugin, and not only do they fix some issues with the latest release of wpgatsby, the schema seems to be resolving more quickly. great work!

**Review by TylerBarnes — APPROVED** [review]
Jason and I just spent 5 hours reviewing this whole thing 😱 and it's ready to merge and release! :D

We're waiting until Monday because we need to make a new source plugin release at the same time since this is a breaking change.

**Comment by acao** [comment]
awesome work! yeah all the tests are passing with this branch against the source plugin primary branch.

File: src/ActionMonitor/Monitors/NavMenuMonitor.php [hunk 1]
```diff
@@ -150,7 +150,7 @@ public function log_diffed_menus( array $old_locations, array $new_locations ) {
 				$this->log_action(
 					[
 						'action_type'         => 'DELETE',
-						'title'               => $removed_menu_id,
+						'title'               => 'Menu: ' . $removed_menu_id,
```
Comment (TylerBarnes) [inline src/ActionMonitor/Monitors/NavMenuMonitor.php:153]: This should be the menu title or just the menu id (`#${removed_menu_id)`)

File: src/ActionMonitor/Monitors/UserMonitor.php [hunk 2]
```diff
@@ -13,12 +13,42 @@ class UserMonitor extends Monitor {
...
+	 */
+	protected $post_ids_to_reassign;
+
 	/**
 	 * Initialize UserMonitor Actions
 	 *
 	 * @return mixed|void
 	 */
-	public function init() {
+	public function init() {if ( ! empty( $post_ids ) && is_array( $post_ids ) )  {
```
Comment (TylerBarnes) [inline src/ActionMonitor/Monitors/UserMonitor.php:27]: Looks like the formatting is a bit weird here. And where does $post_ids come from? Is that a global?

File: src/ActionMonitor/Monitors/UserMonitor.php [hunk 3]
```diff
@@ -135,10 +165,31 @@ public function callback_delete_user( $user_id, $reassign_id ) {
...
+		if ( ! empty( $reassign_user ) ) {
+
+			// @todo: We should get rid of this as it can get expensive to log these actions.
+			// Gatsby Source WordPress should have support for bulk-actions so we can log a single action
+			// such as "DELETE_AUTHOR_AND_REASSIGN_POSTS" and pass the old author ID and new author ID and
+			// Gatsby could do it without an action per modified post.
+			global $wpdb;
+			$post_types = $this->action_monitor->get_tracked_post_types();
+			$post_types = implode( "', '", $post_types );
+			$post_ids = $wpdb->get_col( $wpdb->prepare( "SELECT ID FROM $wpdb->posts WHERE post_author = %d AND post_type IN ('$post_types')", $user_id ) );
```
Comment (TylerBarnes) [inline src/ActionMonitor/Monitors/UserMonitor.php:184]: This needs to be published posts only. Right now it's also creating actions to reassign drafts and trashed posts.

**Review by TylerBarnes — CHANGES_REQUESTED** [review]
Just a couple more changes!

File: src/ActionMonitor/Monitors/UserMonitor.php (hunk 2 above)
Comment (jasonbahl) [reply src/ActionMonitor/Monitors/UserMonitor.php:27]: 😱  this was some funky copy-pasta.

File: src/ActionMonitor/Monitors/UserMonitor.php (hunk 3 above)
Comment (jasonbahl) [reply src/ActionMonitor/Monitors/UserMonitor.php:184]: Fixed. And included a test to ensure this behavior continues.

**Comment by TylerBarnes** [comment]
@jasonbahl is this ready?

**Comment by jasonbahl** [comment]
@TylerBarnes I think so!

**Comment by TylerBarnes** [comment]
Looks and works great :D again, amazing work @jasonbahl , so stoked about this! I'll merge and release shortly

**Pull request merged**